### PR TITLE
[AMDGPU] Add MachineVerifier check to detect illegal copies from vector register to SGPR

### PR DESCRIPTION
**Pull request by AditiRM**
Addition of a check in the MachineVerifier to detect and report illegal vector registers to SGPR copies in the AMDGPU backend, ensuring correct code generation.  
We can enforce this check only after SIFixSGPRCopies pass.
This is half-fix in the pipeline with the help of isSSA MachineFuction property, the check is happening for passes after phi-node-elimination.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 1]
```diff
@@ -4613,15 +4613,41 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
...
-    return true;
-
   const MachineFunction *MF = MI.getParent()->getParent();
   const MachineRegisterInfo &MRI = MF->getRegInfo();
 
+  if (SIInstrInfo::isGenericOpcode(MI.getOpcode())) {
+    // FIXME: At this point the COPY verify is done only for non-ssa forms.
+    // Find a better property to recognize the point where instruction selection
+    // is just done.
+    if (!MRI.isSSA() && MI.isCopy())
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4645]: Why is the isSSA check needed?

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 2]
```diff
@@ -4613,15 +4613,41 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
          SubReg.getReg() == SuperVec.getReg();
 }
 
+// Verify the illgal copy from VGPR to SGPR for generic opcode COPY
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4616]: Typo "illegal"

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 3]
```diff
@@ -4613,15 +4613,41 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
...
+bool SIInstrInfo::verifyCopy(const MachineInstr &MI,
+                             const MachineRegisterInfo &MRI,
+                             StringRef &ErrInfo) const {
+  const MachineOperand &Dst = MI.getOperand(0);
+  const MachineOperand &Src = MI.getOperand(1);
+
+  if (Dst.isReg() && Src.isReg()) {
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4623]: MachineVerifier already enforces that both operands are registers so you should not need to check here.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 4]
```diff
@@ -4613,15 +4613,41 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
...
+    Register DstReg = Dst.getReg();
+    Register SrcReg = Src.getReg();
+    // This is a check for copy from an VGPR to SGPR
+    if (RI.isVGPR(MRI, SrcReg) && RI.isSGPRReg(MRI, DstReg)) {
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4627]: Does this also diagnose AGPR to SGPR copies?

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4645]: Because we lack the appropriate property to express the real condition, and this is a half measure. We really want a custom IsSelectedAndSGPRCopiesFixed property, otherwise the natural check will fail before SIFixSGPRCopies

File: llvm/test/CodeGen/AMDGPU/fix-illegal-vgpr-copies.mir [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# RUN: not --crash llc -march=amdgcn -mcpu=gfx1200 -run-pass=machineverifier %s 2>&1 | FileCheck -check-prefix=ERR %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/fix-illegal-vgpr-copies.mir:1]: This belongs in test/MachineVerifier. Also you can just use -run-pass=none, the verifier runs on the input anyway

File: llvm/test/CodeGen/AMDGPU/fix-illegal-vgpr-copies.mir [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    %4:sgpr_32 = COPY %0:vgpr_32
+
+    ; ERR: *** Bad machine code: illegal copy from VGPR to SGPR ***
+    ; ERR: instruction: $sgpr0 = COPY %0:vgpr_32
+    $sgpr0 = COPY %0:vgpr_32
+    
+    ; ERR: *** Bad machine code: illegal copy from VGPR to SGPR ***
+    ; ERR: instruction: $sgpr1 = COPY $vgpr0
+    $sgpr1 = COPY $vgpr0
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/fix-illegal-vgpr-copies.mir:27]: Test the AGPR case, and test some register tuple classes.

File: llvm/test/CodeGen/AMDGPU/phi-moveimm-subreg-input.mir [hunk 7]
```diff
@@ -0,0 +1,30 @@
+# RUN: not --crash llc  -mtriple=amdgcn -mcpu=gfx900 -verify-machineinstrs -run-pass=si-fix-sgpr-copies -o - %s 2>&1 | FileCheck -check-prefix=GCN %s
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/phi-moveimm-subreg-input.mir:2]: This looks like an unrelated test?

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 1 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4645]: OK, then the comment should say explicitly say that we can only enforce this after SIFixSGPRCopies.

File: llvm/test/CodeGen/AMDGPU/phi-moveimm-subreg-input.mir (hunk 7 above)
Comment (AditiRM) [reply llvm/test/CodeGen/AMDGPU/phi-moveimm-subreg-input.mir:2]: In this specific testcase(phi-vgpr-input-moveimm.mir) it is throwing illegal copy error as expected, and it is because isSSA considers subreg defs invalid in SSA. So I have extracted the specific subtest & converted it into not crash. Please let me know if this is valid.

Reference : https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/MIRParser/MIRParser.cpp#L361

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 4 above)
Comment (cdevadas) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4627]: Also, this only validates in one direction. shouldn't we check the other way as well?
SGPR -> VectorClass
VectorClass -> SGPR

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 1 above)
Comment (cdevadas) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4645]: I was wondering; the existing ` MachineFunctionProperties::Property::Selected` currently set in the GISel pipeline when selection is completed, should be set in the SelectionDAG as well (mostly after finalize-isel) for consistency. MachineFunctions currently don't have this property set for SelectionDAG.
If this field is rightly set, we could catch the illegal copies soon after finalize-isel, by then SIFixSGPRCopies would have been invoked.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 4 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4627]: No! COPY from SGPR to VGPR is fine.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4645]: The problem is if selectiondag sets it, the natural place it would be set in selection is still before SIFixSGPRCopies

Comment (cdevadas) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4645]: Isn't FinalizeISel that marks the end of selection? If not, the pass name is misleading.

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4645]: Yes and no, and I don't think it's really misleading. It finalizes the already selected function into a verifiable form

File: llvm/test/CodeGen/AMDGPU/fix-illegal-vgpr-copies.mir (hunk 6 above)
Comment (AditiRM) [reply llvm/test/CodeGen/AMDGPU/fix-illegal-vgpr-copies.mir]: Updated the testcase with AGPR and tuple cases

File: llvm/test/CodeGen/AMDGPU/fix-illegal-vgpr-copies.mir (hunk 5 above)
Comment (AditiRM) [reply llvm/test/CodeGen/AMDGPU/fix-illegal-vgpr-copies.mir]: Done.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 4 above)
Comment (AditiRM) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp]: Added the illegal copy check from AGPR to SGPR.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 3 above)
Comment (AditiRM) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp]: Done

File: llvm/test/CodeGen/AMDGPU/phi-moveimm-subreg-input.mir (hunk 7 above)
Comment (AditiRM) [reply llvm/test/CodeGen/AMDGPU/phi-moveimm-subreg-input.mir:2]: @arsenm Please take a look at this testcase change.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 8]
```diff
@@ -4613,15 +4613,45 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
...
 
+// Verify the illegal copy from VGPR to SGPR for generic opcode COPY
+bool SIInstrInfo::verifyCopy(const MachineInstr &MI,
+                             const MachineRegisterInfo &MRI,
+                             StringRef &ErrInfo) const {
+  const MachineOperand &Dst = MI.getOperand(0);
+  const MachineOperand &Src = MI.getOperand(1);
+
+  Register DstReg = Dst.getReg();
+  Register SrcReg = Src.getReg();
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4624]:
```suggestion
  Register DstReg = MI.getOperand(0).getReg();
  Register SrcReg = MI.getOperand(1).getReg();
```

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 9]
```diff
@@ -4613,15 +4613,45 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
...
+  if (RI.isVGPR(MRI, SrcReg) && RI.isSGPRReg(MRI, DstReg)) {
+    ErrInfo = "illegal copy from VGPR to SGPR";
+    return false;
+  }
+
+  // This is a check for copy from an AGPR to SGPR
+  if (RI.isAGPR(MRI, SrcReg) && RI.isSGPRReg(MRI, DstReg)) {
+    ErrInfo = "illegal copy from AGPR to SGPR";
+    return false;
+  }
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4635]: I would prefer a single check with RI.isVectorRegister() that includes both VGPRs and AGPRs. Change the error message to "illegal copy from vector register to SGPR"

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 1 above)
Comment (cdevadas) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4650]: Improve the comment to include Jay's suggestion.
"explicitly say that we can only enforce this after SIFixSGPRCopies."

File: llvm/test/CodeGen/AMDGPU/phi-moveimm-subreg-input.mir (hunk 7 above)
Comment (cdevadas) [reply llvm/test/CodeGen/AMDGPU/phi-moveimm-subreg-input.mir:2]: I guess this test should be removed. See my comment below in the original test.

File: llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir [hunk 10]
```diff
@@ -32,38 +32,6 @@ body:             |
     S_BRANCH %bb.2
 ...
 
----
-# GCN-LABEL: name: phi_moveimm_subreg_input
```
Comment (cdevadas) [inline llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:36]: This test doesn't look legitimate to be invoked by a pass that needs SSA from. The presence of the subreg def makes this test a non-SSA one.

**Comment by cdevadas** [comment]
Change the PR heading and the description to include vector register instead of VGPR.

File: llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir (hunk 10 above)
Comment (AditiRM) [reply llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:36]: So should I remove this testcase itself ? Also do we need to add this extracted testcase again to the original one ?

Comment (cdevadas) [reply llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:36]: Yes, avoid this new test case. I recommend deleting this instance from the original test file.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 1 above)
Comment (AditiRM) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4641]: Improved the comment.

File: llvm/test/CodeGen/AMDGPU/phi-moveimm-subreg-input.mir (hunk 7 above)
Comment (AditiRM) [reply llvm/test/CodeGen/AMDGPU/phi-moveimm-subreg-input.mir]: I removed this testcase instance.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 8 above)
Comment (AditiRM) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp]: Done

File: llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir (hunk 10 above)
Comment (AditiRM) [reply llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:36]: Done

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 9 above)
Comment (AditiRM) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp]: Updated the error message with single check.

**Review by cdevadas — APPROVED** [review]
LGTM

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 11]
```diff
@@ -4613,15 +4613,36 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
...
 bool SIInstrInfo::verifyInstruction(const MachineInstr &MI,
                                     StringRef &ErrInfo) const {
   uint16_t Opcode = MI.getOpcode();
-  if (SIInstrInfo::isGenericOpcode(MI.getOpcode()))
-    return true;
-
   const MachineFunction *MF = MI.getParent()->getParent();
   const MachineRegisterInfo &MRI = MF->getRegInfo();
 
+  if (SIInstrInfo::isGenericOpcode(MI.getOpcode())) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4636]: The isGenericOpcode check is redundant with the isCopy check

File: llvm/test/MachineVerifier/fix-illegal-copies.mir [hunk 12]
```diff
@@ -0,0 +1,60 @@
+# RUN: not --crash llc -march=amdgcn -mcpu=gfx1200 -run-pass=none -o /dev/null %s 2>&1 | FileCheck %s    
+# REQUIRES: amdgpu-registered-target
```
Comment (arsenm) [inline llvm/test/MachineVerifier/fix-illegal-copies.mir:2]: Test file name doesn't indicate it's an AMDGPU test. We probably should have an AMDGPU directory at this point too

File: llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir [hunk 13]
```diff
@@ -32,38 +32,6 @@ body:             |
...
-    S_BRANCH %bb.2
-
-  bb.2:
-    successors: %bb.3
-    %3:sreg_64 = PHI %1, %bb.3, %2, %bb.1
-    S_BRANCH %bb.3
-
-  bb.3:
-    successors: %bb.2
-    undef %1.sub0:sreg_64 = COPY %0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:62]: Can you move the sub register indices to the phi input operands, instead of deleting the test

File: llvm/test/MachineVerifier/fix-illegal-copies.mir (hunk 12 above)
Comment (AditiRM) [reply llvm/test/MachineVerifier/fix-illegal-copies.mir]: Implemented this in new PR and updated in latest commit.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 11 above)
Comment (AditiRM) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4631]: Yes, fixed.

File: llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir (hunk 13 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:62]: ping?

Comment (AditiRM) [reply llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:62]: Updated the testcase. Please check @arsenm.

File: llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir [hunk 14]
```diff
@@ -34,36 +34,35 @@ body:             |
...
   bb.1:
     successors: %bb.2
-    undef %2.sub0:sreg_64 = S_ADD_U32 %4, %5, implicit-def $scc
+    %2:sreg_32 = S_ADD_U32 %4, %5, implicit-def $scc
     S_BRANCH %bb.2
 
   bb.2:
     successors: %bb.3
-    %3:sreg_64 = PHI %1, %bb.3, %2, %bb.1
+    %3:sreg_32 = PHI %1, %bb.3, %2, %bb.1
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:57]: %1 should probably be a full copy, and this operand %1.sub0 (same for %2)

**Review by arsenm — APPROVED** [review]

File: llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir (hunk 14 above)
Comment (AditiRM) [reply llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:57]: Can you please explain why do you think we should be introducing subreg in PHI input itself ?

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:57]: When the phi is being processed, the only relevant sub registers would come from the phi's own operands, not some other instructions

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 15]
```diff
@@ -4600,15 +4600,37 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
...
   const MachineRegisterInfo &MRI = MF->getRegInfo();
 
+  // FIXME: At this point the COPY verify is done only for non-ssa forms.
+  // Find a better property to recognize the point where instruction selection
+  // is just done.
+  // We can only enforce this check after SIFixSGPRCopies pass.
+  if (!MRI.isSSA() && MI.isCopy())
+    return verifyCopy(MI, MRI, ErrInfo);
+
+  if (SIInstrInfo::isGenericOpcode(MI.getOpcode())) {
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4630]: No need of braces.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 16]
```diff
@@ -4600,15 +4600,37 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
...
-  if (SIInstrInfo::isGenericOpcode(MI.getOpcode()))
-    return true;
-
   const MachineFunction *MF = MI.getParent()->getParent();
   const MachineRegisterInfo &MRI = MF->getRegInfo();
 
+  // FIXME: At this point the COPY verify is done only for non-ssa forms.
+  // Find a better property to recognize the point where instruction selection
+  // is just done.
+  // We can only enforce this check after SIFixSGPRCopies pass.
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4626]: The last sentence is incomplete. May be, something like the following:
We can only enforce this check after SIFixSGPRCopies so that the illegal copies are legalized and thereafter we don't expect a pass inserting similar copies.

**Comment by arsenm** [comment]
> LLVM Buildbot has detected a new failure on builder `llvm-clang-x86_64-expensive-checks-debian` running on `gribozavr4` while building `llvm` at step 6 "test-build-unified-tree-check-all".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/16/builds/5632
> 
> Here is the relevant piece of the build log for the reference

Fixed f9bd08382a4ad84c06e0b572d3b7fc3ecdb81898
